In [104]:
import pandas as pd
import numpy as np

In [105]:
df = pd.read_csv('./data/Aviation_Data.csv', parse_dates=['Event.Date', 'Publication.Date'], low_memory=False)

#Sam's addition, parsing dates

/var/folders/vf/_ggtcvh11slbnt8cr1ghjwxw0000gn/T/ipykernel_56918/2267368882.py:1: UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df = pd.read_csv('./data/Aviation_Data.csv', parse_dates=['Event.Date', 'Publication.Date'], low_memory=False)


In [230]:
pd.set_option('display.max_columns', None)

In [106]:
df.head()
# something looks off with event date because the years are not in range of data set

Event.Id Investigation.Type Accident.Number Event.Date  \
0  20001218X45444           Accident      SEA87LA080 1948-10-24   
1  20001218X45447           Accident      LAX94LA336 1962-07-19   
2  20061025X01555           Accident      NYC07LA005 1974-08-30   
3  20001218X45448           Accident      LAX96LA321 1977-06-19   
4  20041105X01764           Accident      CHI79FA064 1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaT  
1               UNK                Unknown  Probable Cause       1996-09-19  
2               IMC                 Cruise  Probable Cause       2007-02-26  
3               IMC                 Cruise  Probable Cause       2000-09-12  
4               VMC               Approach  Probable Cause       1980-04-16  

[5 rows x 31 columns]

In [107]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90348 entries, 0 to 90347
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Event.Id                88889 non-null  object        
 1   Investigation.Type      90348 non-null  object        
 2   Accident.Number         88889 non-null  object        
 3   Event.Date              88889 non-null  datetime64[ns]
 4   Location                88837 non-null  object        
 5   Country                 88663 non-null  object        
 6   Latitude                34382 non-null  object        
 7   Longitude               34373 non-null  object        
 8   Airport.Code            50132 non-null  object        
 9   Airport.Name            52704 non-null  object        
 10  Injury.Severity         87889 non-null  object        
 11  Aircraft.damage         85695 non-null  object        
 12  Aircraft.Category       32287 non-null  object

In [108]:
df.describe(include='all')

Event.Id Investigation.Type Accident.Number  \
count            88889              90348           88889   
unique           87951                 71           88863   
top     20001212X19172           Accident      CEN22LA149   
freq                 3              85015               2   
mean               NaN                NaN             NaN   
min                NaN                NaN             NaN   
25%                NaN                NaN             NaN   
50%                NaN                NaN             NaN   
75%                NaN                NaN             NaN   
max                NaN                NaN             NaN   
std                NaN                NaN             NaN   

                           Event.Date       Location        Country Latitude  \
count                           88889          88837          88663    34382   
unique                            NaN          27758            219    25589   
top                               NaN  ANCHORAGE, AK  United States  332739N   
freq                              NaN            434          82248       19   
mean    1999-09-17 17:13:39.354476032            NaN            NaN      NaN   
min               1948-10-24 00:00:00            NaN            NaN      NaN   
25%               1989-01-15 00:00:00            NaN            NaN      NaN   
50%               1998-07-18 00:00:00            NaN            NaN      NaN   
75%               2009-07-01 00:00:00            NaN            NaN      NaN   
max               2022-12-29 00:00:00            NaN            NaN      NaN   
std                               NaN            NaN            NaN      NaN   

       Longitude Airport.Code Airport.Name  ... Purpose.of.flight Air.carrier  \
count      34373        50132        52704  ...             82697       16648   
unique     27154        10374        24870  ...                26       13590   
top     0112457W         NONE      Private  ...          Personal       Pilot   
freq          24         1488          240  ...             49448         258   
mean         NaN          NaN          NaN  ...               NaN         NaN   
min          NaN          NaN          NaN  ...               NaN         NaN   
25%          NaN          NaN          NaN  ...               NaN         NaN   
50%          NaN          NaN          NaN  ...               NaN         NaN   
75%          NaN          NaN          NaN  ...               NaN         NaN   
max          NaN          NaN          NaN  ...               NaN         NaN   
std          NaN          NaN          NaN  ...               NaN         NaN   

       Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
count          77488.000000           76379.000000         76956.000000   
unique                  NaN                    NaN                  NaN   
top                     NaN                    NaN                  NaN   
freq                    NaN                    NaN                  NaN   
mean               0.647855               0.279881             0.357061   
min                0.000000               0.000000             0.000000   
25%                0.000000               0.000000             0.000000   
50%                0.000000               0.000000             0.000000   
75%                0.000000               0.000000             0.000000   
max              349.000000             161.000000           380.000000   
std                5.485960               1.544084             2.235625   

       Total.Uninjured Weather.Condition  Broad.phase.of.flight  \
count     82977.000000             84397                  61724   
unique             NaN                 4                     12   
top                NaN               VMC                Landing   
freq               NaN             77303                  15428   
mean          5.325440               NaN                    NaN   
min           0.000000               NaN    

In [109]:
df['Investigation.Type'].value_counts()

Investigation.Type
Accident      85015
Incident       3874
25-09-2020      702
26-09-2020       60
02-02-2021       39
              ...  
13-09-2021        1
04-08-2021        1
05-08-2022        1
03-11-2020        1
06-01-2021        1
Name: count, Length: 71, dtype: int64

In [110]:
df['Purpose.of.flight'].value_counts()
# filter values for private vs commercial

Purpose.of.flight
Personal                     49448
Instructional                10601
Unknown                       6802
Aerial Application            4712
Business                      4018
Positioning                   1646
Other Work Use                1264
Ferry                          812
Aerial Observation             794
Public Aircraft                720
Executive/corporate            553
Flight Test                    405
Skydiving                      182
External Load                  123
Public Aircraft - Federal      105
Banner Tow                     101
Air Race show                   99
Public Aircraft - Local         74
Public Aircraft - State         64
Air Race/show                   59
Glider Tow                      53
Firefighting                    40
Air Drop                        11
ASHO                             6
PUBS                             4
PUBL                             1
Name: count, dtype: int64

In [111]:
df['Air.carrier'].value_counts()

Air.carrier
Pilot                        258
American Airlines             90
United Airlines               89
Delta Air Lines               53
SOUTHWEST AIRLINES CO         42
                            ... 
WOODY CONTRACTING INC          1
Rod Aviation LLC               1
Paul D Franzon                 1
TRAINING SERVICES INC DBA      1
MC CESSNA 210N LLC             1
Name: count, Length: 13590, dtype: int64

In [112]:
df['Location'].value_counts()
#filter data for country and state using .str.split()

Location
ANCHORAGE, AK          434
MIAMI, FL              200
ALBUQUERQUE, NM        196
HOUSTON, TX            193
CHICAGO, IL            184
                      ... 
Corona De Tucso, AZ      1
Lithonia, GA             1
BONANZA, OR              1
NEWPORT, PA              1
Brasnorte,               1
Name: count, Length: 27758, dtype: int64

In [113]:
df['Total.Uninjured'].value_counts()

Total.Uninjured
0.0      29879
1.0      25101
2.0      15988
3.0       4313
4.0       2662
         ...  
558.0        1
412.0        1
338.0        1
401.0        1
455.0        1
Name: count, Length: 379, dtype: int64

# Sam's Addition

In [114]:
df.head(20)

Event.Id Investigation.Type Accident.Number Event.Date  \
0   20001218X45444           Accident      SEA87LA080 1948-10-24   
1   20001218X45447           Accident      LAX94LA336 1962-07-19   
2   20061025X01555           Accident      NYC07LA005 1974-08-30   
3   20001218X45448           Accident      LAX96LA321 1977-06-19   
4   20041105X01764           Accident      CHI79FA064 1979-08-02   
5   20170710X52551           Accident      NYC79AA106 1979-09-17   
6   20001218X45446           Accident      CHI81LA106 1981-08-01   
7   20020909X01562           Accident      SEA82DA022 1982-01-01   
8   20020909X01561           Accident      NYC82DA015 1982-01-01   
9   20020909X01560           Accident      MIA82DA029 1982-01-01   
10  20020909X01559           Accident      FTW82DA034 1982-01-01   
11  20020909X01558           Accident      ATL82DKJ10 1982-01-01   
12  20020917X02148           Accident      FTW82FRJ07 1982-01-02   
13  20020917X02134           Accident      FTW82FRA14 1982-01-02   
14  20020917X02119           Accident      FTW82FPJ10 1982-01-02   
15  20020917X02117           Accident      FTW82FPG08 1982-01-02   
16  20020917X01962           Accident      DEN82DTM08 1982-01-02   
17  20020917X01656           Accident      ANC82FAG14 1982-01-02   
18  20020917X02481           Accident      NYC82DA016 1982-01-02   
19  20020917X02339           Accident      MIA82DA028 1982-01-02   

            Location        Country   Latitude   Longitude Airport.Code  \
0    MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1     BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2      Saltville, VA  United States  36.922223  -81.878056          NaN   
3         EUREKA, CA  United States        NaN         NaN          NaN   
4         Canton, OH  United States        NaN         NaN          NaN   
5         BOSTON, MA  United States  42.445277  -70.758333          NaN   
6         COTTON, MN  United States        NaN         NaN          NaN   
7        PULLMAN, WA  United States        NaN         NaN          NaN   
8   EAST HANOVER, NJ  United States        NaN         NaN          N58   
9   JACKSONVILLE, FL  United States        NaN         NaN          JAX   
10         HOBBS, NM  United States        NaN         NaN          NaN   
11      TUSKEGEE, AL  United States        NaN         NaN          NaN   
12         HOMER, LA  United States        NaN         NaN          NaN   
13        HEARNE, TX  United States        NaN         NaN          T72   
14     CHICKASHA, OK  United States        NaN         NaN          NaN   
15   LITTLE ROCK, AR  United States        NaN         NaN          NaN   
16        MIDWAY, UT  United States        NaN         NaN          NaN   
17       SKWENTA, AK  United States        NaN         NaN          NaN   
18       GALETON, PA  United States        NaN         NaN          5G6   
19         MIAMI, FL  United States        NaN         NaN          NaN   

          Airport.Name  ... Purpose.of.flight Air.carrier  \
0                  NaN  ...          Personal         NaN   
1                  NaN  ...          Personal         NaN   
2                  NaN  ...          Personal         NaN   
3                  NaN  ...          Personal         NaN   
4                  NaN  ...          Personal         NaN   
5                  NaN  ...               NaN  Air Canada   
6                  NaN  ...          Personal         NaN   
7   BLACKBURN AG STRIP  ...          Personal         NaN   
8              HANOVER  ...          Business         NaN   
9    JACKSONVILLE INTL  ...          Personal         NaN   
10                 NaN  ...          Personal         NaN   
11            TUSKEGEE  ...          Personal         NaN   
12                 NaN  ...          Personal         NaN   
13    HEARNE MUNICIPAL  ...          Personal         NaN   
14                 NaN  ...          Personal         NaN   
15                 NaN  ...  

We want to string split so that we can lump by states and countries

In [115]:
df['Location_City'] = df['Location'].str.split(", ").str[0].str.upper()

In [116]:
df['Location_State_Country'] = df['Location'].str.split(", ").str[1].str.upper()

In [117]:
df['Location_State_Country'].value_counts().head()

Location_State_Country
CA    8857
TX    5913
FL    5825
AK    5672
AZ    2834
Name: count, dtype: int64

In [118]:
df.head()

Event.Id Investigation.Type Accident.Number Event.Date  \
0  20001218X45444           Accident      SEA87LA080 1948-10-24   
1  20001218X45447           Accident      LAX94LA336 1962-07-19   
2  20061025X01555           Accident      NYC07LA005 1974-08-30   
3  20001218X45448           Accident      LAX96LA321 1977-06-19   
4  20041105X01764           Accident      CHI79FA064 1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Total.Fatal.Injuries Total.Serious.Injuries  \
0          NaN  ...                  2.0                    0.0   
1          NaN  ...                  4.0                    0.0   
2          NaN  ...                  3.0                    NaN   
3          NaN  ...                  2.0                    0.0   
4          NaN  ...                  1.0                    2.0   

  Total.Minor.Injuries Total.Uninjured Weather.Condition  \
0                  0.0             0.0               UNK   
1                  0.0             0.0               UNK   
2                  NaN             NaN               IMC   
3                  0.0             0.0               IMC   
4                  NaN             0.0               VMC   

  Broad.phase.of.flight   Report.Status  Publication.Date Location_City  \
0                Cruise  Probable Cause               NaT   MOOSE CREEK   
1               Unknown  Probable Cause        1996-09-19    BRIDGEPORT   
2                Cruise  Probable Cause        2007-02-26     SALTVILLE   
3                Cruise  Probable Cause        2000-09-12        EUREKA   
4              Approach  Probable Cause        1980-04-16        CANTON   

  Location_State_Country  
0                     ID  
1                     CA  
2                     VA  
3                     CA  
4                     OH  

[5 rows x 33 columns]

In [119]:
df.tail()

Event.Id Investigation.Type Accident.Number Event.Date  \
90343  20221227106491           Accident      ERA23LA093 2022-12-26   
90344  20221227106494           Accident      ERA23LA095 2022-12-26   
90345  20221227106497           Accident      WPR23LA075 2022-12-26   
90346  20221227106498           Accident      WPR23LA076 2022-12-26   
90347  20221230106513           Accident      ERA23LA097 2022-12-29   

            Location        Country Latitude Longitude Airport.Code  \
90343  Annapolis, MD  United States      NaN       NaN          NaN   
90344    Hampton, NH  United States      NaN       NaN          NaN   
90345     Payson, AZ  United States  341525N  1112021W          PAN   
90346     Morgan, UT  United States      NaN       NaN          NaN   
90347     Athens, GA  United States      NaN       NaN          NaN   

      Airport.Name  ... Total.Fatal.Injuries Total.Serious.Injuries  \
90343          NaN  ...                  0.0                    1.0   
90344          NaN  ...                  0.0                    0.0   
90345       PAYSON  ...                  0.0                    0.0   
90346          NaN  ...                  0.0                    0.0   
90347          NaN  ...                  0.0                    1.0   

      Total.Minor.Injuries Total.Uninjured Weather.Condition  \
90343                  0.0             0.0               NaN   
90344                  0.0             0.0               NaN   
90345                  0.0             1.0               VMC   
90346                  0.0             0.0               NaN   
90347                  0.0             1.0               NaN   

      Broad.phase.of.flight Report.Status  Publication.Date Location_City  \
90343                   NaN           NaN        2022-12-29     ANNAPOLIS   
90344                   NaN           NaN               NaT       HAMPTON   
90345                   NaN           NaN        2022-12-27        PAYSON   
90346                   NaN           NaN               NaT        MORGAN   
90347                   NaN           NaN        2022-12-30        ATHENS   

      Location_State_Country  
90343                     MD  
90344                     NH  
90345                     AZ  
90346                     UT  
90347                     GA  

[5 rows x 33 columns]

After investigating the 'Investigation.Type' column, we have identified there should be only 2 categories ("Accident vs. Incident"), summing to a total of 88,889 rows vs. the total 90,348. We have looked at '25-09-2020', '26-09-2020', and '02-02-2021', all of which show rows consisting of null values. We need to filter out the dataset to show only the 2 categories.

In [120]:
df[df['Investigation.Type'] ==  '02-02-2021'].head()

Event.Id Investigation.Type Accident.Number Event.Date Location Country  \
75894      NaN         02-02-2021             NaN        NaT      NaN     NaN   
78228      NaN         02-02-2021             NaN        NaT      NaN     NaN   
78351      NaN         02-02-2021             NaN        NaT      NaN     NaN   
78499      NaN         02-02-2021             NaN        NaT      NaN     NaN   
78513      NaN         02-02-2021             NaN        NaT      NaN     NaN   

      Latitude Longitude Airport.Code Airport.Name  ... Total.Fatal.Injuries  \
75894      NaN       NaN          NaN          NaN  ...                  NaN   
78228      NaN       NaN          NaN          NaN  ...                  NaN   
78351      NaN       NaN          NaN          NaN  ...                  NaN   
78499      NaN       NaN          NaN          NaN  ...                  NaN   
78513      NaN       NaN          NaN          NaN  ...                  NaN   

      Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
75894                    NaN                  NaN             NaN   
78228                    NaN                  NaN             NaN   
78351                    NaN                  NaN             NaN   
78499                    NaN                  NaN             NaN   
78513                    NaN                  NaN             NaN   

      Weather.Condition Broad.phase.of.flight Report.Status  Publication.Date  \
75894               NaN                   NaN           NaN               NaT   
78228               NaN                   NaN           NaN               NaT   
78351               NaN                   NaN           NaN               NaT   
78499               NaN                   NaN           NaN               NaT   
78513               NaN                   NaN           NaN               NaT   

      Location_City Location_State_Country  
75894           NaN                    NaN  
78228           NaN                    NaN  
78351           NaN                    NaN  
78499           NaN                    NaN  
78513           NaN                    NaN  

[5 rows x 33 columns]

In [121]:
df = df[(df['Investigation.Type'] == 'Accident') | (df['Investigation.Type'] == 'Incident')]

In [122]:
#checking to see if df was filtered down to just Accidents and Incidents

df['Investigation.Type'].value_counts()

Investigation.Type
Accident    85015
Incident     3874
Name: count, dtype: int64

Next we want to investigate the date range earlier than 1962 because it seems like there are dates before this year.

In [123]:
df[(df['Event.Date'] < '1962-01-01')]

Event.Id Investigation.Type Accident.Number Event.Date  \
0  20001218X45444           Accident      SEA87LA080 1948-10-24   

          Location        Country Latitude Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States      NaN       NaN          NaN   

  Airport.Name  ... Total.Fatal.Injuries Total.Serious.Injuries  \
0          NaN  ...                  2.0                    0.0   

  Total.Minor.Injuries Total.Uninjured Weather.Condition  \
0                  0.0             0.0               UNK   

  Broad.phase.of.flight   Report.Status  Publication.Date Location_City  \
0                Cruise  Probable Cause               NaT   MOOSE CREEK   

  Location_State_Country  
0                     ID  

[1 rows x 33 columns]

There seems to be only one date before 1962. We should be able to throw out. However, the Aviation Accident Database provides this acknowledgement.

"Generally, a preliminary report is available online within a few days of an accident. Factual information is added when available, and when the investigation is completed, the preliminary report is replaced with a final description of the accident and its probable cause. Full narrative descriptions may not be available for dates before 1993, cases under revision, or where NTSB did not have primary investigative responsibility."

In [124]:
#We narrow the scope of our data to 1993 and after to start with data with full narrative descriptions

df = df[(df['Event.Date'] >= '1993-01-01')]

In [125]:
df.head()

Event.Id Investigation.Type Accident.Number Event.Date  \
32026  20001211X11729           Accident      MIA93FA041 1993-01-01   
32027  20001211X11711           Accident      LAX93LA087 1993-01-01   
32028  20001211X11645           Accident      ATL93LA045 1993-01-01   
32029  20001211X11732           Accident      MIA93LA042 1993-01-01   
32030  20001211X11647           Accident      ATL93LA048 1993-01-01   

                Location        Country Latitude Longitude Airport.Code  \
32026        LILLIAN, AL  United States      NaN       NaN          NaN   
32027         FOLSOM, CA  United States      NaN       NaN          NaN   
32028  FUQUAY-VARINA, NC  United States      NaN       NaN          5W5   
32029     TITUSVILLE, FL  United States      NaN       NaN          TIX   
32030       BESSEMER, AL  United States      NaN       NaN          2A3   

                 Airport.Name  ... Total.Fatal.Injuries  \
32026                     NaN  ...                  1.0   
32027                     NaN  ...                  0.0   
32028        TRIPLE W AIRPARK  ...                  0.0   
32029  SPACE CENTER EXECUTIVE  ...                  0.0   
32030                BESSEMER  ...                  0.0   

      Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
32026                    0.0                  1.0             0.0   
32027                    1.0                  0.0             0.0   
32028                    0.0                  1.0             0.0   
32029                    0.0                  1.0             1.0   
32030                    0.0                  0.0             2.0   

      Weather.Condition Broad.phase.of.flight   Report.Status  \
32026               VMC              Approach  Probable Cause   
32027               VMC           Maneuvering  Probable Cause   
32028               VMC              Approach  Probable Cause   
32029               VMC              Approach  Probable Cause   
32030               VMC               Takeoff  Probable Cause   

       Publication.Date  Location_City Location_State_Country  
32026        1993-11-15        LILLIAN                     AL  
32027        1993-10-13         FOLSOM                     CA  
32028        1993-10-25  FUQUAY-VARINA                     NC  
32029        1994-02-10     TITUSVILLE                     FL  
32030        1993-11-03       BESSEMER                     AL  

[5 rows x 33 columns]

In [126]:
#We take a look at where the null values are

df.isna().sum()

Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     41
Country                      73
Latitude                  22490
Longitude                 22499
Airport.Code              23560
Airport.Name              22780
Injury.Severity            1000
Aircraft.damage            2498
Aircraft.Category         28248
Registration.Number        1360
Make                         56
Model                        72
Amateur.Built               101
Number.of.Engines          5749
Engine.Type                7092
FAR.Description           28512
Schedule                  49496
Purpose.of.flight          6145
Air.carrier               41966
Total.Fatal.Injuries      11245
Total.Serious.Injuries    12309
Total.Minor.Injuries      11726
Total.Uninjured            5774
Weather.Condition          4489
Broad.phase.of.flight     26851
Report.Status              6384
Publication.Date           2784
Location

In [127]:
# We still have more cleaning up to do. Let's look at the country column
df['Country'].value_counts()

Country
United States                       50738
Brazil                                373
Canada                                350
United Kingdom                        342
Mexico                                334
                                    ...  
Saint Vincent and the Grenadines        1
Belarus                                 1
Malampa                                 1
AY                                      1
Turks and Caicos Islands                1
Name: count, Length: 213, dtype: int64

In [128]:
#Let's look at Weather Conditions
df['Weather.Condition'].value_counts()

Weather.Condition
VMC    48399
IMC     3335
UNK      378
Unk      262
Name: count, dtype: int64

There are 4 weather conditions (https://support.crewlounge.aero/support/solutions/articles/24000034167-difference-between-vfr-ifr-vmc-imc-and-actual-simulated-simulator-time)

VMC - Visual Meteorologic Condition: Visibility and clouds are clear and allow you to fly the aircraft by visual reference
IMC - Instrument Meteorologic Condition: The weather is restricting your visibility, you need aircraft flight instruments to fly safe

UNK and Unk means unknown. We should consolidate this into 1 grouping

In [129]:
df[df['Weather.Condition'] == "Unk"] = "UNK"

/var/folders/vf/_ggtcvh11slbnt8cr1ghjwxw0000gn/T/ipykernel_56918/4015243441.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'UNK' has dtype incompatible with datetime64[ns], please explicitly cast to a compatible dtype first.
  df[df['Weather.Condition'] == "Unk"] = "UNK"
/var/folders/vf/_ggtcvh11slbnt8cr1ghjwxw0000gn/T/ipykernel_56918/4015243441.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'UNK' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df[df['Weather.Condition'] == "Unk"] = "UNK"


In [130]:
#Check the Weather Condition Categories
df['Weather.Condition'].value_counts()

Weather.Condition
VMC    48399
IMC     3335
UNK      640
Name: count, dtype: int64

Let's check for duplicate Accident Numbers

In [131]:
#Total Number of Accident Unfiltered is 56863
len(df['Accident.Number'])

56863

In [132]:
#Check for Duplicates
df['Accident.Number'].value_counts().head()

Accident.Number
UNK           262
WPR22LA143      2
DCA22WA130      2
DCA22LA201      2
GAA22WA241      2
Name: count, dtype: int64

In [133]:
df[df['Accident.Number'] == 'UNK']

Event.Id Investigation.Type Accident.Number Event.Date Location Country  \
67173      UNK                UNK             UNK        UNK      UNK     UNK   
67763      UNK                UNK             UNK        UNK      UNK     UNK   
68347      UNK                UNK             UNK        UNK      UNK     UNK   
69021      UNK                UNK             UNK        UNK      UNK     UNK   
69509      UNK                UNK             UNK        UNK      UNK     UNK   
...        ...                ...             ...        ...      ...     ...   
90019      UNK                UNK             UNK        UNK      UNK     UNK   
90135      UNK                UNK             UNK        UNK      UNK     UNK   
90140      UNK                UNK             UNK        UNK      UNK     UNK   
90171      UNK                UNK             UNK        UNK      UNK     UNK   
90219      UNK                UNK             UNK        UNK      UNK     UNK   

      Latitude Longitude Airport.Code Airport.Name  ... Total.Fatal.Injuries  \
67173      UNK       UNK          UNK          UNK  ...                  UNK   
67763      UNK       UNK          UNK          UNK  ...                  UNK   
68347      UNK       UNK          UNK          UNK  ...                  UNK   
69021      UNK       UNK          UNK          UNK  ...                  UNK   
69509      UNK       UNK          UNK          UNK  ...                  UNK   
...        ...       ...          ...          ...  ...                  ...   
90019      UNK       UNK          UNK          UNK  ...                  UNK   
90135      UNK       UNK          UNK          UNK  ...                  UNK   
90140      UNK       UNK          UNK          UNK  ...                  UNK   
90171      UNK       UNK          UNK          UNK  ...                  UNK   
90219      UNK       UNK          UNK          UNK  ...                  UNK   

      Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
67173                    UNK                  UNK             UNK   
67763                    UNK                  UNK             UNK   
68347                    UNK                  UNK             UNK   
69021                    UNK                  UNK             UNK   
69509                    UNK                  UNK             UNK   
...                      ...                  ...             ...   
90019                    UNK                  UNK             UNK   
90135                    UNK                  UNK             UNK   
90140                    UNK                  UNK             UNK   
90171                    UNK                  UNK             UNK   
90219                    UNK                  UNK             UNK   

      Weather.Condition Broad.phase.of.flight Report.Status Publication.Date  \
67173               UNK                   UNK           UNK              UNK   
67763               UNK                   UNK           UNK              UNK   
68347               UNK                   UNK           UNK              UNK   
69021               UNK                   UNK           UNK              UNK   
69509               UNK                   UNK           UNK              UNK   
...                 ...                   ...           ...              ...   
90019               UNK                   UNK           UNK              UNK   
90135               UNK                   UNK           UNK              UNK   
90140               UNK                   UNK           UNK              UNK   
90171               UNK                   UNK           UNK              UNK   
90219               UNK                   UNK           UNK              UNK   

      Location_City Location_State_Country  
67173           UNK                    UNK  
67763           UNK                    UNK  
68347           UNK                    UNK  
69021           UNK                    UNK  
69509           UNK                    UNK  
...         

In [134]:
df[df['Accident.Number'] == 'ERA22FA318']

Event.Id Investigation.Type Accident.Number           Event.Date  \
89593  20220718105496           Accident      ERA22FA318  2022-07-17 00:00:00   
89594  20220718105496           Accident      ERA22FA318  2022-07-17 00:00:00   

            Location        Country Latitude Longitude Airport.Code  \
89593  Las Vegas, NV  United States  361239N  1151140W          VGT   
89594  Las Vegas, NV  United States  361239N  1151140W          VGT   

          Airport.Name  ... Total.Fatal.Injuries Total.Serious.Injuries  \
89593  NORTH LAS VEGAS  ...                  4.0                    0.0   
89594  NORTH LAS VEGAS  ...                  4.0                    0.0   

      Total.Minor.Injuries Total.Uninjured Weather.Condition  \
89593                  0.0             0.0               VMC   
89594                  0.0             0.0               VMC   

      Broad.phase.of.flight Report.Status     Publication.Date Location_City  \
89593                   NaN           NaN  2022-08-05 00:00:00     LAS VEGAS   
89594                   NaN           NaN  2022-08-05 00:00:00     LAS VEGAS   

      Location_State_Country  
89593                     NV  
89594                     NV  

[2 rows x 33 columns]

In [135]:
df[df['Accident.Number'] == 'DCA22WA167']

Event.Id Investigation.Type Accident.Number           Event.Date  \
88619  20220801105632           Incident      DCA22WA167  2022-01-22 00:00:00   
88620  20220801105632           Incident      DCA22WA167  2022-01-22 00:00:00   

       Location Country Latitude Longitude Airport.Code Airport.Name  ...  \
88619  Kigali,   Rwanda  001587S  0030822E          NaN          NaN  ...   
88620  Kigali,   Rwanda  001587S  0030822E          NaN          NaN  ...   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
88619                  0.0                    0.0                  0.0   
88620                  0.0                    0.0                  0.0   

      Total.Uninjured Weather.Condition Broad.phase.of.flight Report.Status  \
88619             0.0               NaN                   NaN           NaN   
88620             0.0               NaN                   NaN           NaN   

      Publication.Date Location_City Location_State_Country  
88619              NaT        KIGALI                         
88620              NaT        KIGALI                         

[2 rows x 33 columns]

Based on the investigation, we can remove the 'UNK' category in the Accident.Number, then we will need to drop duplicates

In [136]:
#First drop the UNK category

df = df[df['Accident.Number'] != 'UNK']

In [137]:
df['Accident.Number'].value_counts()

Accident.Number
CEN22FA424    2
DCA22WA172    2
WPR23LA041    2
DCA22WA130    2
WPR22FA309    2
             ..
LAX01LA224    1
FTW01LA206    1
CHI01LA191    1
ATL01LA069    1
ERA23LA097    1
Name: count, Length: 56575, dtype: int64

In [138]:
#Next drop duplicates
df = df.drop_duplicates(subset=['Accident.Number'])

In [139]:
df['Accident.Number'].value_counts()

Accident.Number
MIA93FA041    1
ERA11CA226    1
WPR11LA179    1
WPR11LA180    1
WPR11LA181    1
             ..
LAX01LA224    1
FTW01LA206    1
CHI01LA191    1
ATL01LA069    1
ERA23LA097    1
Name: count, Length: 56575, dtype: int64

We will now look at other categories

In [140]:
df['Aircraft.Category'].value_counts()

#From looking at the categories - We can narrow the focus to just Airplane

Aircraft.Category
Airplane             24097
Helicopter            3105
Glider                 453
Balloon                200
Weight-Shift           161
Gyrocraft              158
Powered Parachute       91
Ultralight              29
Unknown                 13
WSFT                     9
Powered-Lift             5
Blimp                    4
UNK                      2
Rocket                   1
ULTR                     1
Name: count, dtype: int64

In [141]:
df['Make'].value_counts()

Make
Cessna                        10732
Piper                          5787
CESSNA                         4867
PIPER                          2817
Beech                          2209
                              ...  
Zivko Aeronautics Inc.            1
Mil Design Bureau                 1
Scandinavian Aero Industry        1
Amateur Built                     1
ROYSE RALPH L                     1
Name: count, Length: 6875, dtype: int64

In [142]:
df['Model'].value_counts()

Model
172                     1118
152                     1030
172N                     685
PA-28-140                490
737                      473
                        ... 
Marquart ChargerMA-5       1
EMB-135-KL                 1
LG2H                       1
Trainer                    1
M-8 EAGLE                  1
Name: count, Length: 9701, dtype: int64

In [143]:
df['Amateur.Built'].value_counts()

#Given that we are addressing commercial and private enterprises, we want to remove Amateur Built Aircrafts

Amateur.Built
No     50266
Yes     6208
Name: count, dtype: int64

We narrow down the df to just Airplanes and Not Amateur Built

In [144]:
df = df[(df['Aircraft.Category'] == 'Airplane') & (df['Amateur.Built'] == 'No')]

In [145]:
#Checking Aircraft Category
df['Aircraft.Category'].value_counts()

Aircraft.Category
Airplane    21115
Name: count, dtype: int64

In [146]:
df['Amateur.Built'].value_counts()

Amateur.Built
No    21115
Name: count, dtype: int64

We will now look at other categories

In [147]:
df['FAR.Description'].value_counts()

#FAR means Federal Aviation Regulations (https://www.ecfr.gov/current/title-14/chapter-I/subchapter-G)


FAR.Description
091                              13340
Part 91: General Aviation         2459
NUSN                              1069
NUSC                               779
137                                741
121                                602
135                                535
UNK                                221
129                                194
Part 137: Agricultural             150
PUBU                               124
Part 121: Air Carrier              106
Part 135: Air Taxi & Commuter       71
Part 129: Foreign                   71
Non-U.S., Non-Commercial            71
Non-U.S., Commercial                67
091K                                14
Public Use                          12
Unknown                             12
125                                  5
ARMF                                 4
Part 125: 20+ Pax,6000+ lbs          2
Part 91 Subpart K: Fractional        1
107                                  1
Name: count, dtype: int64

List of FAA Acronym List (https://www.proairpilot.com/faa-acronyms-list.html)

FAR means Federal Aviation Regulations (https://www.ecfr.gov/current/title-14/chapter-I/subchapter-G)
- Part 91: General Operating and Flight Rules
- Part 91/Subpart K: Fractional Ownership Operations
- Part 107: Small Unmanned Aircraft Systems
- Part 121: Operating Requirements: Domestic, Flag, and Supplemental Operations (Air Carrier)
- Part 125: Certification and Operations: Aircraft Having a Seating Capacity of 20 or More Passengers or a Maximum Payload Capacity of 6,000 Pounds or More; and Rules Governing Persons on Board Such Aircraft (20+ Pax, 6000+)
- Part 129: Operations: Foreign Air Carriers and Foreign Operators of U.S. - Registered Aircraft Engaged in Common Carriage
- Part 135: Operating Requirements: Commuter and on Demand Operations and Rules Governing Persons on Board Such Aircraft (Air Taxi & Commuter)
- Part 137: Agricultural Aircraft Operations
- Non-U.S., Non-Commercial
- Non-U.S., Commercial
- ARMF
- NUSN
- NUSC
- PUBU
- Public Use
- UNK

This one might be harder to categorize into smaller groups (Hard to understand what some of these acronyms mean)

In [148]:
df['Schedule'].value_counts()

Schedule
SCHD    1311
NSCH    1063
UNK       30
Name: count, dtype: int64

In [149]:
df['Purpose.of.flight'].value_counts()

Purpose.of.flight
Personal                     11620
Instructional                 2736
Aerial Application             929
Business                       489
Positioning                    343
Unknown                        332
Aerial Observation             159
Other Work Use                 152
Skydiving                      150
Flight Test                    120
Ferry                          102
Executive/corporate            102
Banner Tow                      89
Public Aircraft - Federal       51
Air Race show                   48
Public Aircraft                 37
Glider Tow                      35
Public Aircraft - State         24
Firefighting                    16
Public Aircraft - Local         12
Air Race/show                    4
ASHO                             4
Air Drop                         3
PUBS                             3
External Load                    1
Name: count, dtype: int64

What some of the Purpose of flights mean (chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/https://www.ntsb.gov/safety/safety-studies/Documents/SS0101.pdf) or 
chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/https://www.faa.gov/sites/faa.gov/files/2022GASurvey-Appendix-B-Documents_04DEC2023V1.pdf or chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/https://ia903102.us.archive.org/17/items/6560171-Survival-Flight-Accident-Report/6560171-Survival-Flight-Accident-Report.pdf:

Purpose of Flight seems like a good category to reclassify in new groupings. Just to clean up the data, it might be worth reclassifying the following to the 15 categories (16 with the Unknown Category) in the Table 4 for Purpose of Flight Categories

Things to change:
- Positioning to "Other"
- Skydiving is under 14 CFR Part 105 Parachuting guidelines, which should be listed under "Other Work Use" (which includes parachuting)
- Flight Test to "Other" (Proficiency Flight is Flight Test)
- Ferry to "Other"
- Banner Tow to "Other Work Use"
- "Public Use": Public Aircraft, Public Aircraft - Federal, Public Aircraft - State, Public Aircraft - Local
- Air Race show & Air Race/show to Air Race
- Glider Tow to "Other Work Use"
- Firefighting to "Other Aerial Application"
- ASHO to "Air Race"
- PUBS to "Public Use"
- Air Drop to "Other Work Use"

# 03/04/24

In [150]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21115 entries, 32592 to 90345
Data columns (total 33 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Event.Id                21115 non-null  object
 1   Investigation.Type      21115 non-null  object
 2   Accident.Number         21115 non-null  object
 3   Event.Date              21115 non-null  object
 4   Location                21109 non-null  object
 5   Country                 21114 non-null  object
 6   Latitude                18971 non-null  object
 7   Longitude               18965 non-null  object
 8   Airport.Code            13836 non-null  object
 9   Airport.Name            13923 non-null  object
 10  Injury.Severity         20332 non-null  object
 11  Aircraft.damage         19952 non-null  object
 12  Aircraft.Category       21115 non-null  object
 13  Registration.Number     20910 non-null  object
 14  Make                    21112 non-null  object
 15  Mod

Up to this point we have filtered for Investigation.Type, Accident.Number, Event.Date, Location(City,StateCountry), Aircraft.Category, Amateur.Built, and Weather.Condition. 

In [151]:
df.head(50)

Event.Id Investigation.Type Accident.Number           Event.Date  \
32592  20001211X12054           Accident      ANC93FA054  1993-04-23 00:00:00   
33535  20001211X13214           Accident      NYC93FA154  1993-08-14 00:00:00   
33580  20001211X13180           Accident      LAX93LA329  1993-08-18 00:00:00   
33711  20001211X13402           Accident      FTW93LA257  1993-09-04 00:00:00   
33880  20001211X13354           Accident      BFO93LA184  1993-09-25 00:00:00   
34143  20001211X13771           Accident      LAX94WA053  1993-11-13 00:00:00   
34210  20001211X13695           Accident      BFO94FA021  1993-11-28 00:00:00   
34275  20001211X13814           Accident      ANC94LA026  1993-12-12 00:00:00   
34342  20001206X00618           Accident     FTW94FA060B  1994-01-02 00:00:00   
34557  20001206X00742           Accident      DCA94WA036  1994-02-22 00:00:00   
35545  20001206X01822           Accident      NYC94FA128  1994-07-16 00:00:00   
35887  20001206X02019           Accident      CHI94LA298  1994-08-25 00:00:00   
35933  20001206X01926           Accident      ANC94LA126  1994-08-31 00:00:00   
36047  20001206X02310           Accident      LAX94WA373  1994-09-14 00:00:00   
36095  20001206X02325           Accident      MIA94WA215  1994-09-20 00:00:00   
36113  20001206X02234           Accident      DCA94WA077  1994-09-23 00:00:00   
36401  20001206X02621           Accident      LAX95WA056  1994-11-18 00:00:00   
36569  20001206X02737           Accident      LAX95FA062  1994-12-27 00:00:00   
36581  20001206X02711           Accident      DCA95WA012  1994-12-29 00:00:00   
36882  20001207X03179           Accident      NYC95LA070  1995-03-11 00:00:00   
36893  20001207X03162           Accident     MIA95LA086A  1995-03-12 00:00:00   
36894  20001207X03162           Accident     MIA95LA086B  1995-03-12 00:00:00   
37150  20001207X03370           Accident      NYC95LA093  1995-04-24 00:00:00   
37171  20001207X03355           Accident      MIA95RA121  1995-04-28 00:00:00   
37214  20001207X03480           Accident      FTW95WA227  1995-05-06 00:00:00   
37280  20001207X03526           Incident     NYC95IA114B  1995-05-19 00:00:00   
37281  20001207X03526           Incident     NYC95IA114A  1995-05-19 00:00:00   
37845  20001207X03963           Accident      FTW95LA309  1995-07-23 00:00:00   
38293  20001207X04585           Accident      NYC95MA220  1995-09-10 00:00:00   
38336  20001207X04563           Accident      MIA95WA232  1995-09-15 00:00:00   
38337  20001207X04527           Accident      LAX95LA335  1995-09-15 00:00:00   
38640  20001207X04862           Incident      FTW96RA039  1995-11-07 00:00:00   
38846  20001207X04990           Accident      DCA96RA020  1995-12-20 00:00:00   
39039  20001208X05308           Accident      NYC96LA058  1996-02-07 00:00:00   
39137  20010304X00536           Accident     DCA96RA042B  1996-02-24 00:00:00   
39138  20010304X00536           Accident     DCA96RA042A  1996-02-24 00:00:00   
39223  20001208X05343           Accident      ANC96LA033  1996-03-13 00:00:00   
39393  20001208X05582           Accident      IAD96FA061  1996-04-14 00:00:00   
39476  20001208X05646           Accident      MIA96LA135  1996-04-27 00:00:00   
40009  20001208X06203           Accident      DCA96MA068  1996-07-06 00:00:00   
40280  20001208X06417           Accident      ATL96LA117  1996-08-07 00:00:00   
40288  20001208X06448           Accident      CHI96LA281  1996-08-07 00:00:00   
40824  20001208X06865           Accident      DCA97WA004  1996-10-31 00:00:00   
40856  20001208X07016           Incident      DCA97WA005  1996-11-06 00:00:00   
40930  20001208X07017           Accident      DCA97WA011  1996-11-23 00:00:00   
41259  20010711X01386           Incident      DCA97WA029  1997-02-15 00:00:00   
41340  20001208X07516           Accident      DCA97WA032  1997-03-07 00:00:00   
41555  20001208X07696           Accident      DCA97WA039  1997-04-10 00:00:00   
41610  20001208X07660           Accident    

In [152]:
df['Report.Status'].value_counts()

Report.Status
Probable Cause                                                                                                                                                                                                                                                                  2739
Foreign                                                                                                                                                                                                                                                                          283
<br /><br />                                                                                                                                                                                                                                                                     125
The pilot's failure to maintain directional control during the landing roll.                                                                               

In [153]:
df['Injury.Severity'].value_counts()

Injury.Severity
Non-Fatal      16179
Fatal           3499
Minor            145
Serious          107
Incident         104
Fatal(1)          96
Fatal(2)          74
Fatal(3)          26
Fatal(4)          24
Unavailable       16
Fatal(5)          15
Fatal(6)          11
Fatal(7)           6
Fatal(8)           5
Fatal(10)          4
Fatal(13)          2
Fatal(154)         2
Fatal(12)          2
Fatal(65)          1
Fatal(160)         1
Fatal(97)          1
Fatal(113)         1
Fatal(26)          1
Fatal(19)          1
Fatal(9)           1
Fatal(54)          1
Fatal(24)          1
Fatal(125)         1
Fatal(131)         1
Fatal(169)         1
Fatal(18)          1
Fatal(228)         1
Fatal(20)          1
Name: count, dtype: int64

In [154]:
df['Injury.Severity'] = df['Injury.Severity'].map(lambda x: 'Fatal' if isinstance(x, str) and 'Fatal(' in x else x)

In [155]:
df['Injury.Severity']

32592        Fatal
33535        Fatal
33580    Non-Fatal
33711        Fatal
33880    Non-Fatal
           ...    
90328    Non-Fatal
90332    Non-Fatal
90335    Non-Fatal
90336        Minor
90345    Non-Fatal
Name: Injury.Severity, Length: 21115, dtype: object

In [156]:
df[df['Injury.Severity'] == 'Unavailable']

Event.Id Investigation.Type Accident.Number           Event.Date  \
41656  20001208X07697           Accident      DCA97WA043  1997-04-27 00:00:00   
42287  20001208X08352           Accident      DCA97WA054  1997-07-21 00:00:00   
42847  20001208X08992           Accident      DCA98WA082  1997-10-09 00:00:00   
44014  20001211X10045           Accident      DCA98WA048  1998-05-29 00:00:00   
44608  20001211X10826           Accident      DCA98WA079  1998-08-05 00:00:00   
45357  20001211X11538           Accident      DCA99WA020  1998-12-05 00:00:00   
45560  20001204X00045           Accident      DCA99WA035  1999-01-28 00:00:00   
45711  20001205X00299           Accident      DCA99WA050  1999-03-04 00:00:00   
45764  20001205X00297           Accident      DCA99WA048  1999-03-15 00:00:00   
45978  20001205X00483           Accident      IAD99WA071  1999-04-24 00:00:00   
46927  20010330X00670           Accident      DCA99WA091  1999-08-21 00:00:00   
47191  20001212X19734           Accident      DCA99WA086  1999-09-23 00:00:00   
48024  20001212X20608           Accident      DCA00WA039  2000-03-20 00:00:00   
48384  20010105X00025           Accident      DCA00WA050  2000-05-17 00:00:00   
49914  20010329X00665           Accident      DCA01WA012  2001-01-06 00:00:00   
60181  20081003X63354           Accident      ENG06RA013  2006-02-05 00:00:00   

                           Location             Country Latitude Longitude  \
41656                  JINAN, CHINA               China      NaN       NaN   
42287     SOUTHERN THAILA, Thailand            Thailand      NaN       NaN   
42847                 HARBIN, CHINA               China      NaN       NaN   
44014                 BOMBAY, INDIA               India      NaN       NaN   
44608     SEOUL, KOREA, REPUBLIC OF  Korea, Republic Of      NaN       NaN   
45357          MANILLA, Philippines         Philippines      NaN       NaN   
45560                 NAPLES, ITALY               Italy      NaN       NaN   
45711              BIARRITZ, FRANCE              France      NaN       NaN   
45764    POHANG, KOREA, REPUBLIC OF  Korea, Republic Of      NaN       NaN   
45978            Georgetown, Guyana              Guyana      NaN       NaN   
46927              ISTANBUL, Turkey              Turkey      NaN       NaN   
47191             BANGKOK, THAILAND            Thailand      NaN       NaN   
48024                AREQUIPA, PERU                Peru      NaN       NaN   
48384                       , Gabon               Gabon      NaN       NaN   
49914                 Dundo, Angola              Angola      NaN       NaN   
60181  STRAHAN, TASMANIA, AUSTRALIA           Australia      NaN       NaN   

      Airport.Code Airport.Name  ... Total.Fatal.Injuries  \
41656          NaN          NaN  ...                  NaN   
42287          NaN          NaN  ...                  NaN   
42847          NaN          NaN  ...                  NaN   
44014          NaN          NaN  ...                  NaN   
44608          NaN          NaN  ...                  NaN   
45357          NaN          NaN  ...                  NaN   
45560          NaN          NaN  ...                  NaN   
45711          NaN          NaN  ...                  NaN   
45764          NaN          NaN  ...                  NaN   
45978          NaN          NaN  ...                  NaN   
46927          NaN          NaN  ...                  NaN   
47191          NaN          NaN  ...                  NaN   
48024          NaN          NaN  ...                  NaN   
48384          NaN          NaN  ...                  NaN   
49914          NaN          NaN  ...                  NaN   
60181          NaN          NaN  ...                  NaN   

      Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
41656                    NaN                  NaN             NaN   
42287                    NaN                  NaN             NaN   
42847                    NaN                  NaN             N

In [157]:
df[df['Country'] == 'Brazil']

Event.Id Investigation.Type Accident.Number           Event.Date  \
40824  20001208X06865           Accident      DCA97WA004  1996-10-31 00:00:00   
44286  20001211X10547           Incident      DCA98WA070  1998-07-02 00:00:00   
44359  20001211X10551           Accident      DCA98WA078  1998-07-10 00:00:00   
45449  20001211X11543           Incident      DCA99WA027  1998-12-28 00:00:00   
47911  20001212X20476           Accident      DCA00WA029  2000-02-27 00:00:00   
...               ...                ...             ...                  ...   
89180  20220531105151           Accident      GAA22WA190  2022-05-17 00:00:00   
89367  20220622105302           Accident      GAA22WA208  2022-06-15 00:00:00   
89423  20220701105398           Accident      GAA22WA214  2022-06-22 00:00:00   
89954  20220914105933           Accident      GAA22WA281  2022-09-09 00:00:00   
89959  20220913105922           Accident      GAA22WA278  2022-09-10 00:00:00   

                        Location Country Latitude Longitude Airport.Code  \
40824          SAO PAULO, Brazil  Brazil      NaN       NaN          NaN   
44286             BRAZIL, Brazil  Brazil      NaN       NaN          NaN   
44359          SAO PAULO, BRAZIL  Brazil      NaN       NaN          NaN   
45449            CURITBA, BRAZIL  Brazil      NaN       NaN          NaN   
47911       PORTO ALEGRE, BRAZIL  Brazil      NaN       NaN          NaN   
...                          ...     ...      ...       ...          ...   
89180   Brasilandia De Minas, OF  Brazil  001715S  0046040W          NaN   
89367                 Itaituba,   Brazil  051850S  0573933W          NaN   
89423                Benevides,   Brazil  011846S  0481956W         SNYP   
89954  Santa Helena de Goias, OF  Brazil  174816S  0050294W          NaN   
89959              Oriximina, OF  Brazil      NaN       NaN          NaN   

      Airport.Name  ... Total.Fatal.Injuries Total.Serious.Injuries  \
40824          NaN  ...                 97.0                    4.0   
44286          NaN  ...                  NaN                    NaN   
44359          NaN  ...                  NaN                    NaN   
45449          NaN  ...                  NaN                    NaN   
47911          NaN  ...                  NaN                    NaN   
...            ...  ...                  ...                    ...   
89180          NaN  ...                  1.0                    0.0   
89367          NaN  ...                  1.0                    0.0   
89423          NaN  ...                  0.0                    0.0   
89954          NaN  ...                  0.0                    0.0   
89959          NaN  ...                  1.0                    5.0   

      Total.Minor.Injuries Total.Uninjured Weather.Condition  \
40824                  2.0             NaN               VMC   
44286                  NaN             5.0               UNK   
44359                  NaN             5.0               UNK   
45449                  NaN            40.0               UNK   
47911                  NaN           123.0               IMC   
...                    ...             ...               ...   
89180                  0.0             0.0               NaN   
89367                  0.0             0.0               NaN   
89423                  0.0             2.0               NaN   
89954                  0.0             0.0               NaN   
89959                  0.0             0.0               NaN   

      Broad.phase.of.flight Report.Status     Publication.Date  \
40824                   NaN       Foreign                  NaT   
44286                   NaN       Foreign                  NaT   
44359                   NaN       Foreign                  NaT   
45449                   NaN       Foreign                  NaT   
47911                   NaN       Foreign                  NaT   
...                     ...           ...                  ...   
89180                   NaN           NaN  

After evaluation of the Countries we have determined keeping international countries will be difficult given the limited data we have. As such we are sticking to just the U.S. 

In [159]:
df['Country'].value_counts()

Country
United States               17957
Brazil                        221
United Kingdom                199
Mexico                        199
Canada                        161
                            ...  
Cayenne                         1
Solomon Islands                 1
Qatar                           1
Singapore                       1
Turks and Caicos Islands        1
Name: count, Length: 161, dtype: int64

In [160]:
df = df[df['Country'] == 'United States']

In [161]:
df['Country'].value_counts()

Country
United States    17957
Name: count, dtype: int64

In [162]:
df.isna().sum()

Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                      0
Country                       0
Latitude                    135
Longitude                   142
Airport.Code               4832
Airport.Name               4769
Injury.Severity              68
Aircraft.damage             400
Aircraft.Category             0
Registration.Number          10
Make                          0
Model                         6
Amateur.Built                 0
Number.of.Engines           472
Engine.Type                1623
FAR.Description               2
Schedule                  16489
Purpose.of.flight          1168
Air.carrier                9403
Total.Fatal.Injuries       2527
Total.Serious.Injuries     2504
Total.Minor.Injuries       2221
Total.Uninjured             510
Weather.Condition           426
Broad.phase.of.flight     15224
Report.Status              1899
Publication.Date           1286
Location

In [163]:
df['Schedule'].value_counts()

Schedule
NSCH    820
SCHD    643
UNK       5
Name: count, dtype: int64

We feel comfortable at this point identifying columns to drop before we continue filtering. The ones we want to drop are Latitude, longitude, Broad phase of flight, Schedule. We feel they are not relevant or have to many missing values and we have deemed them not very informative for analysis purposes.

In [164]:
df = df.drop(columns= ['Latitude', 'Longitude', 'Broad.phase.of.flight', 'Schedule'])

In [165]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17957 entries, 32592 to 90345
Data columns (total 29 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Event.Id                17957 non-null  object
 1   Investigation.Type      17957 non-null  object
 2   Accident.Number         17957 non-null  object
 3   Event.Date              17957 non-null  object
 4   Location                17957 non-null  object
 5   Country                 17957 non-null  object
 6   Airport.Code            13125 non-null  object
 7   Airport.Name            13188 non-null  object
 8   Injury.Severity         17889 non-null  object
 9   Aircraft.damage         17557 non-null  object
 10  Aircraft.Category       17957 non-null  object
 11  Registration.Number     17947 non-null  object
 12  Make                    17957 non-null  object
 13  Model                   17951 non-null  object
 14  Amateur.Built           17957 non-null  object
 15  Num

In [62]:
df.isna().sum()

Event.Id                     0
Investigation.Type           0
Accident.Number              0
Event.Date                   0
Location                     0
Country                      0
Airport.Code              4832
Airport.Name              4769
Injury.Severity             68
Aircraft.damage            400
Aircraft.Category            0
Registration.Number         10
Make                         0
Model                        6
Amateur.Built                0
Number.of.Engines          472
Engine.Type               1623
FAR.Description              2
Purpose.of.flight         1168
Air.carrier               9403
Total.Fatal.Injuries      2527
Total.Serious.Injuries    2504
Total.Minor.Injuries      2221
Total.Uninjured            510
Weather.Condition          426
Report.Status             1899
Publication.Date          1286
Location_City                0
Location_State_Country      43
dtype: int64

In [166]:
df['Airport.Name'].value_counts()

Airport.Name
Private                           147
Private Airstrip                   91
PVT                                23
Private Strip                      19
Merrill Field                      18
                                 ... 
Delphi Muni                         1
Willow Pond Aviation Inc            1
Brownsville / S Padre Island        1
Ridgley Airpark                     1
WICHITA DWIGHT D EISENHOWER NT      1
Name: count, Length: 8226, dtype: int64

We have done our second round of observing null values and having investigated Airport.Code, Airport.Name, and Air.carrier we have determined that these values are also unnecessary for our analysis. 

In [167]:
df = df.drop(columns= ['Airport.Code', 'Airport.Name', 'Air.carrier'])

In [168]:
df.isna().sum()

Event.Id                     0
Investigation.Type           0
Accident.Number              0
Event.Date                   0
Location                     0
Country                      0
Injury.Severity             68
Aircraft.damage            400
Aircraft.Category            0
Registration.Number         10
Make                         0
Model                        6
Amateur.Built                0
Number.of.Engines          472
Engine.Type               1623
FAR.Description              2
Purpose.of.flight         1168
Total.Fatal.Injuries      2527
Total.Serious.Injuries    2504
Total.Minor.Injuries      2221
Total.Uninjured            510
Weather.Condition          426
Report.Status             1899
Publication.Date          1286
Location_City                0
Location_State_Country      43
dtype: int64

In [66]:
df['Report.Status'].value_counts()

Report.Status
Probable Cause                                                                                                                                                                                                                                                                                                                    2733
The pilot's failure to maintain directional control during the landing roll.                                                                                                                                                                                                                                                        52
A loss of engine power for undetermined reasons.                                                                                                                                                                                                                                                                                    45
The p

In [169]:
df['Report.Status'] = df['Report.Status'].fillna(value= 'Unknown')

In [170]:
df.isna().sum()

Event.Id                     0
Investigation.Type           0
Accident.Number              0
Event.Date                   0
Location                     0
Country                      0
Injury.Severity             68
Aircraft.damage            400
Aircraft.Category            0
Registration.Number         10
Make                         0
Model                        6
Amateur.Built                0
Number.of.Engines          472
Engine.Type               1623
FAR.Description              2
Purpose.of.flight         1168
Total.Fatal.Injuries      2527
Total.Serious.Injuries    2504
Total.Minor.Injuries      2221
Total.Uninjured            510
Weather.Condition          426
Report.Status                0
Publication.Date          1286
Location_City                0
Location_State_Country      43
dtype: int64

In [171]:
df['Purpose.of.flight'].value_counts()

Purpose.of.flight
Personal                     11351
Instructional                 2680
Aerial Application             879
Business                       454
Positioning                    330
Aerial Observation             145
Skydiving                      142
Other Work Use                 139
Flight Test                    114
Executive/corporate             93
Banner Tow                      89
Ferry                           87
Unknown                         69
Public Aircraft - Federal       47
Air Race show                   39
Glider Tow                      34
Public Aircraft                 33
Public Aircraft - State         24
Firefighting                    14
Public Aircraft - Local         12
Air Race/show                    4
ASHO                             4
Air Drop                         3
PUBS                             3
Name: count, dtype: int64

We observed that Publication.Date isn't useful for our analysis.

In [173]:
POF_dict = {"Positioning": "Other",
            "Skydiving": "Other Work Use", 
            "Flight Test": "Other",
            "Ferry": "Other", 
            "Banner Tow": "Other Work Use", 
            "Public Aircraft": "Public Use", 
            "Public Aircraft - Federal": "Public Use",
            "Public Aircraft - State": "Public Use", 
            "Public Aircraft - Local": "Public Use",
            "PUBS": "Public Use",
            "Air Race show": "Air Race",
            "Air Race/show": "Air Race",
            "ASHO": "Air Race",
            "Glider Tow": "Other Work Use",
            "Firefighting": "Other Aerial Application",
            "Air Drop": "Other Work Use"}
POF_dict


{'Positioning': 'Other',
 'Skydiving': 'Other Work Use',
 'Flight Test': 'Other',
 'Ferry': 'Other',
 'Banner Tow': 'Other Work Use',
 'Public Aircraft': 'Public Use',
 'Public Aircraft - Federal': 'Public Use',
 'Public Aircraft - State': 'Public Use',
 'Public Aircraft - Local': 'Public Use',
 'PUBS': 'Public Use',
 'Air Race show': 'Air Race',
 'Air Race/show': 'Air Race',
 'ASHO': 'Air Race',
 'Glider Tow': 'Other Work Use',
 'Firefighting': 'Other Aerial Application',
 'Air Drop': 'Other Work Use'}

In [174]:
#for keys in POF_dict:
    #purposeoff(df, POF_dict[keys], keys)

for key, value in POF_dict.items():
    purposeoff(df, value, key)

In [175]:
df['Purpose.of.flight'].value_counts()

Purpose.of.flight
Personal                    11351
Instructional                2680
Aerial Application            879
Other                         531
Business                      454
Other Work Use                407
Aerial Observation            145
Public Use                    119
Executive/corporate            93
Unknown                        69
Air Race                       47
Other Aerial Application       14
Name: count, dtype: int64

In [186]:
df['Purpose.of.flight'] = df['Purpose.of.flight'].fillna(value= 'Unknown')

In [187]:
df['Purpose.of.flight'].value_counts()

Purpose.of.flight
Personal                    11351
Instructional                2680
Unknown                      1237
Aerial Application            879
Other                         531
Business                      454
Other Work Use                407
Aerial Observation            145
Public Use                    119
Executive/corporate            93
Air Race                       47
Other Aerial Application       14
Name: count, dtype: int64

In [188]:
df.isna().sum()

Event.Id                     0
Investigation.Type           0
Accident.Number              0
Event.Date                   0
Location                     0
Country                      0
Injury.Severity             68
Aircraft.damage            400
Aircraft.Category            0
Registration.Number         10
Make                         0
Model                        6
Amateur.Built                0
Number.of.Engines          472
Engine.Type               1623
FAR.Description              2
Purpose.of.flight            0
Total.Fatal.Injuries      2527
Total.Serious.Injuries    2504
Total.Minor.Injuries      2221
Total.Uninjured            510
Weather.Condition          426
Report.Status                0
Publication.Date          1286
Location_City                0
Location_State_Country      43
dtype: int64

In [194]:
df['FAR.Description'].value_counts()

FAR.Description
091                              13271
Part 91: General Aviation         2431
137                                738
135                                523
121                                481
Part 137: Agricultural             150
PUBU                               123
Part 135: Air Taxi & Commuter       69
Part 121: Air Carrier               61
129                                 39
091K                                14
NUSC                                11
NUSN                                10
Public Use                           9
Part 129: Foreign                    8
UNK                                  5
Unknown                              4
125                                  4
ARMF                                 2
Part 125: 20+ Pax,6000+ lbs          1
Part 91 Subpart K: Fractional        1
Name: count, dtype: int64

In [195]:
df = df.drop(columns= ['Engine.Type', 'Registration.Number', 'FAR.Description', 'Publication.Date', 'Number.of.Engines'])

In [197]:
df.isna().sum()

Event.Id                     0
Investigation.Type           0
Accident.Number              0
Event.Date                   0
Location                     0
Country                      0
Injury.Severity             68
Aircraft.damage            400
Aircraft.Category            0
Make                         0
Model                        6
Amateur.Built                0
Purpose.of.flight            0
Total.Fatal.Injuries      2527
Total.Serious.Injuries    2504
Total.Minor.Injuries      2221
Total.Uninjured            510
Weather.Condition          426
Report.Status                0
Location_City                0
Location_State_Country      43
dtype: int64

In [203]:
df['Injury.Severity'].value_counts()

Injury.Severity
Non-Fatal    15041
Fatal         2575
Minor          137
Serious         94
Incident        42
Name: count, dtype: int64

In [212]:
df['Injury.Severity'] = df['Injury.Severity'].fillna(value= 'Unknown')

In [213]:
df.isna().sum()

Event.Id                     0
Investigation.Type           0
Accident.Number              0
Event.Date                   0
Location                     0
Country                      0
Injury.Severity              0
Aircraft.damage            400
Aircraft.Category            0
Make                         0
Model                        6
Amateur.Built                0
Purpose.of.flight            0
Total.Fatal.Injuries      2527
Total.Serious.Injuries    2504
Total.Minor.Injuries      2221
Total.Uninjured            510
Weather.Condition          426
Report.Status                0
Location_City                0
Location_State_Country      43
dtype: int64

In [214]:
df['Aircraft.damage'].value_counts()

Aircraft.damage
Substantial    15744
Destroyed       1424
Minor            349
Unknown           40
Name: count, dtype: int64

In [219]:
df['Aircraft.damage'] = df['Aircraft.damage'].fillna(value= 'Unknown')

In [220]:
df.isna().sum()

Event.Id                     0
Investigation.Type           0
Accident.Number              0
Event.Date                   0
Location                     0
Country                      0
Injury.Severity              0
Aircraft.damage              0
Aircraft.Category            0
Make                         0
Model                        6
Amateur.Built                0
Purpose.of.flight            0
Total.Fatal.Injuries      2527
Total.Serious.Injuries    2504
Total.Minor.Injuries      2221
Total.Uninjured            510
Weather.Condition          426
Report.Status                0
Location_City                0
Location_State_Country      43
dtype: int64

In [221]:
df['Weather.Condition'].value_counts()

Weather.Condition
VMC    16582
IMC      947
UNK        2
Name: count, dtype: int64

In [222]:
df['Weather.Condition'] = df['Weather.Condition'].fillna(value= 'UNK')

In [223]:
df.isna().sum()

Event.Id                     0
Investigation.Type           0
Accident.Number              0
Event.Date                   0
Location                     0
Country                      0
Injury.Severity              0
Aircraft.damage              0
Aircraft.Category            0
Make                         0
Model                        6
Amateur.Built                0
Purpose.of.flight            0
Total.Fatal.Injuries      2527
Total.Serious.Injuries    2504
Total.Minor.Injuries      2221
Total.Uninjured            510
Weather.Condition            0
Report.Status                0
Location_City                0
Location_State_Country      43
dtype: int64

In [224]:
df['Location_State_Country'].value_counts()

Location_State_Country
CA              1666
TX              1423
AK              1381
FL              1328
AZ               677
CO               558
GA               546
WA               521
NC               426
ID               416
IL               400
NY               399
MI               399
OH               386
OR               363
PA               351
VA               329
MO               319
WI               317
AR               308
MN               304
UT               303
NV               278
IN               269
TN               268
NJ               256
MT               254
OK               253
AL               248
KS               248
NM               244
SC               225
LA               221
MD               206
MA               180
IA               173
MS               172
NE               150
WY               147
KY               138
ME               125
ND               122
SD               112
CT               101
HI                79
WV                73
NH         

In [226]:
df['Location_State_Country'] = df['Location_State_Country'].fillna(value= 'Ocean')

In [227]:
df.isna().sum()

Event.Id                     0
Investigation.Type           0
Accident.Number              0
Event.Date                   0
Location                     0
Country                      0
Injury.Severity              0
Aircraft.damage              0
Aircraft.Category            0
Make                         0
Model                        6
Amateur.Built                0
Purpose.of.flight            0
Total.Fatal.Injuries      2527
Total.Serious.Injuries    2504
Total.Minor.Injuries      2221
Total.Uninjured            510
Weather.Condition            0
Report.Status                0
Location_City                0
Location_State_Country       0
dtype: int64

In [232]:
df.dropna(subset = ['Model'], inplace=True)

In [233]:
df.isna().sum()

Event.Id                     0
Investigation.Type           0
Accident.Number              0
Event.Date                   0
Location                     0
Country                      0
Injury.Severity              0
Aircraft.damage              0
Aircraft.Category            0
Make                         0
Model                        0
Amateur.Built                0
Purpose.of.flight            0
Total.Fatal.Injuries      2522
Total.Serious.Injuries    2498
Total.Minor.Injuries      2217
Total.Uninjured            507
Weather.Condition            0
Report.Status                0
Location_City                0
Location_State_Country       0
dtype: int64

We have filtered all columns except for the 4 "Injuries" columns.

In [237]:
df['Total.Fatal.Injuries'].value_counts()

Total.Fatal.Injuries
0.0      12865
1.0       1359
2.0        754
3.0        221
4.0        132
5.0         48
6.0         22
7.0          8
10.0         6
9.0          4
8.0          3
228.0        1
65.0         1
12.0         1
13.0         1
49.0         1
14.0         1
11.0         1
Name: count, dtype: int64

In [253]:
df[df[['Total.Serious.Injuries', 'Total.Fatal.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured']].isnull().all(axis=1)]

Event.Id Investigation.Type Accident.Number           Event.Date  \
46835  20001212X19419           Accident      ATL99LA117  1999-08-13 00:00:00   
47320  20001212X20015           Accident      MIA00LA012  1999-10-16 00:00:00   
63646  20081219X65251           Incident      ENG08IA001  2007-10-28 00:00:00   

                  Location        Country Injury.Severity Aircraft.damage  \
46835     ORMOND BEACH, FL  United States       Non-Fatal     Substantial   
47320       INDIANTOWN, FL  United States       Non-Fatal     Substantial   
63646  Charlotte (CLT), NC  United States        Incident           Minor   

      Aircraft.Category         Make    Model Amateur.Built Purpose.of.flight  \
46835          Airplane  Quicksilver   GT-500            No           Unknown   
47320          Airplane        Beech      D35            No           Unknown   
63646          Airplane       Boeing  757-225            No           Unknown   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
46835                  NaN                    NaN                  NaN   
47320                  NaN                    NaN                  NaN   
63646                  NaN                    NaN                  NaN   

      Total.Uninjured Weather.Condition   Report.Status    Location_City  \
46835             NaN               UNK  Probable Cause     ORMOND BEACH   
47320             NaN               UNK  Probable Cause       INDIANTOWN   
63646             NaN               UNK  Probable Cause  CHARLOTTE (CLT)   

      Location_State_Country  
46835                     FL  
47320                     FL  
63646                     NC

We have investigated the 4 columns of injuries when they are all null values. The two accidents look like they are thefts and the incident look like a malfunction before departure. All three cases look to have no injuries and we can impute the values as 0.0. 

Given that the rest of the four columns have a value in one of the four columns, we can impute 0.0 for the null values in these columns. 

In [255]:
df = df.fillna(0)

/var/folders/vf/_ggtcvh11slbnt8cr1ghjwxw0000gn/T/ipykernel_56918/3475559988.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df = df.fillna(0)


In [257]:
df.isna().sum()

Event.Id                  0
Investigation.Type        0
Accident.Number           0
Event.Date                0
Location                  0
Country                   0
Injury.Severity           0
Aircraft.damage           0
Aircraft.Category         0
Make                      0
Model                     0
Amateur.Built             0
Purpose.of.flight         0
Total.Fatal.Injuries      0
Total.Serious.Injuries    0
Total.Minor.Injuries      0
Total.Uninjured           0
Weather.Condition         0
Report.Status             0
Location_City             0
Location_State_Country    0
dtype: int64